lms/services/canvas.py

```diff
@@ -1,41 +1,198 @@
-            course_id
+        :param resource_link_id: The id of the assignment
+        :param check_in_course: Attempt to map detect and map the file if it's
+            not in the course before attempting to retrieve it. This is more
+            thorough, but causes an extra API request
         :return: A URL suitable for public presentation of the file
-        :raise  CanvasFileNotFoundInCourse: if check_in_course=True and file_id isn't in course_id
-            be in the course but isn't.
+        :raise CanvasFileNotFoundInCourse: if the file is not accessible and
+            cannot be mapped to a file in the course
+        :raise CanvasAPIResourceNotFound: if the file is not valid at all, and
+            cannot be mapped to a file in the course
+        :raise CanvasAPIPermissionError: If we cannot retrieve the file for
+            any other reason
         """
+        # Check to see if we've mapped this file id to something else before
+        file_mapping = self._file_mapping(resource_link_id)
+        mapped_id = file_mapping.get(file_id, default=file_id)
+
+        # If we are told to check up front, try and fix the file if it's not in
+        # the current course (as far as we can tell from this users point of
+        # view)
+        already_mapped = False
+        if check_in_course and not self._is_file_in_course(mapped_id, course_id):
+            mapped_id = self._match_file_in_course(file_id, course_id, file_mapping)
+            already_mapped = True
 
-        if check_in_course:
-            if not self._file_in_course(file_id, course_id):
+            if not mapped_id:
                 raise CanvasFileNotFoundInCourse(file_id)
 
-        return self.api.public_url(file_id)
+        try:
+            return self.api.public_url(mapped_id)
+
+        except CanvasAPIPermissionError:
+            # If we've already mapped this once, we aren't going to get a
+            # better answer now. So bail out
+            if already_mapped:
+                raise
+
+            if not (
+                mapped_id := self._match_file_in_course(
+                    file_id, course_id, file_mapping
+                )
+            ):
+                # We don't know why this failed, could be a course copy, could
+                # just be a permissions issue, all we know is we can't fix it
+                raise
+
+        return self.api.public_url(mapped_id)
+
+    @lru_cache
+    def _file_mapping(self, resource_link_id):
+        assignment = self._assignment_service.get(
+            tool_consumer_instance_guid=self._application_instance.tool_consumer_instance_guid,
+            resource_link_id=resource_link_id,
+        )
+        return CanvasFileMapping(assignment.extra)
+
+    @cached_property
+    def _application_instance(self):
+        return self._application_instance_service.get()
+
+    @lru_cache
+    def _files_in_course(self, course_id):
+        return self.api.list_files(course_id)
```

Comment:
We cache the file results here because we might use them for the "in the course" check and also the mapping. This would save us an API call in that case.